File: packages/python-sdk/src/vizier/circuit_breaker.py
from __future__ import annotations
import collections
import dataclasses
import time
from typing import Any, Dict, Optional, Tuple

from .canonical import sha256_canonical_json

class CircuitTrippedError(Exception):
    """
    Raised when the Agent Circuit Breaker trips due to an infinite loop,
    runaway tool-calling retry storm, or exceeded session budget.
    """

    def __init__(self, code: str, message: str, details: Optional[Dict[str, Any]] = None):
        super().__init__(f"[{code}] {message}")
        self.code = code
        self.message = message
        self.details = details or {}

@dataclasses.dataclass
class CircuitStatus:
    """Status report returned by CircuitBreaker evaluation."""
    tripped: bool
    code: Optional[str] = None
    message: Optional[str] = None
    consecutive_repeats: int = 0
    session_calls_count: int = 0

class CircuitBreaker:
    """
    Agent Circuit Breaker & Loop Killer.
    Protects against infinite tool-calling loops, runaway LLM retries,
    and unexpected API/LLM token bills.
    """

    def __init__(
        self,
        max_repeated_calls: int = 3,
        time_window_seconds: float = 30.0,
        max_session_calls: int = 25,
        cool_off_seconds: float = 60.0,
        max_session_actions: Optional[int] = None,
    ):
        self.max_repeated_calls = max_repeated_calls
        self.time_window_seconds = time_window_seconds
        self.max_session_calls = max_session_actions if max_session_actions is not None else max_session_calls
        self.cool_off_seconds = cool_off_seconds

        self._history: Dict[str, collections.deque[Tuple[float, str]]] = collections.defaultdict(
            lambda: collections.deque(maxlen=100)
        )
        self._session_totals: Dict[str, int] = collections.defaultdict(int)
        self._tripped_sessions: Dict[str, Tuple[float, str, str]] = {}

    def _signature(self, action_type: str, target: str, parameters: Dict[str, Any]) -> str:
        param_hash = sha256_canonical_json(parameters)
        return f"{action_type}:{target}:{param_hash}"

    def check_and_record(
        self,
        action_type: str,
        target: str,
        parameters: Dict[str, Any],
        session_id: str = "default_session",
    ) -> CircuitStatus:
        now = time.time()

        # 1. Check if session is currently in active cool-off
        if session_id in self._tripped_sessions:
            trip_time, code, msg = self._tripped_sessions[session_id]
            if now - trip_time < self.cool_off_seconds:
                remaining = int(self.cool_off_seconds - (now - trip_time))
                return CircuitStatus(
                    tripped=True,
                    code=code,
                    message=f"Circuit breaker active for session '{session_id}'. Cool-off remaining: {remaining}s. ({msg})",
                    session_calls_count=self._session_totals[session_id],
                )
            else:
                del self._tripped_sessions[session_id]

        # 2. Check total session action budget
        total = self._session_totals[session_id] + 1
        if total > self.max_session_calls:
            code = "CIRCUIT_TRIPPED:BUDGET_EXCEEDED"
            msg = (
                f"Session action budget exceeded (limit: {self.max_session_calls} actions). "
                f"Execution halted to prevent runaway costs."
            )
            self._tripped_sessions[session_id] = (now, code, msg)
            return CircuitStatus(tripped=True, code=code, message=msg, session_calls_count=total)

        # 3. Check repeated identical calls within sliding time window
        sig = self._signature(action_type, target, parameters)
        history = self._history[session_id]

        cutoff = now - self.time_window_seconds
        while history and history[0][0] < cutoff:
            history.popleft()

        consecutive = 0
        for _, prev_sig in reversed(history):
            if prev_sig == sig:
                consecutive += 1
            else:
                break

        history.append((now, sig))
        self._session_totals[session_id] = total

        if consecutive + 1 >= self.max_repeated_calls:
            code = "CIRCUIT_TRIPPED:LOOP_DETECTED"
            msg = (
                f"Potential infinite tool-calling loop detected: action '{action_type}' "
                f"called {consecutive + 1} times with identical parameters within {int(self.time_window_seconds)}s. "
                f"Execution halted."
            )
            self._tripped_sessions[session_id] = (now, code, msg)
            return CircuitStatus(
                tripped=True,
                code=code,
                message=msg,
                consecutive_repeats=consecutive + 1,
                session_calls_count=total,
            )

        return CircuitStatus(
            tripped=False,
            consecutive_repeats=consecutive + 1,
            session_calls_count=total,
        )

    def reset(self, session_id: Optional[str] = None) -> None:
        if session_id is None:
            self._history.clear()
            self._session_totals.clear()
            self._tripped_sessions.clear()
        else:
            self._history.pop(session_id, None)
            self._session_totals.pop(session_id, None)
            self._tripped_sessions.pop(session_id, None)

File: src/core/circuit-breaker.ts
import type { KVNamespace } from "@cloudflare/workers-types";
import { canonicalize, sha256 } from "./receipts";
import type { VerificationRequest } from "./schemas";

export const DEFAULT_MAX_REPEATED_CALLS = 5;
export const DEFAULT_TIME_WINDOW_SECONDS = 30;
export const DEFAULT_COOL_OFF_SECONDS = 60;
export const DEFAULT_MAX_SESSION_ACTIONS = 100;

export interface EdgeCircuitBreakerResult {
  readonly tripped: boolean;
  readonly reasonCode?: "CIRCUIT_TRIPPED:LOOP_DETECTED" | "CIRCUIT_TRIPPED:BUDGET_EXCEEDED";
  readonly message?: string;
  readonly details?: Readonly<Record<string, unknown>>;
}

export interface EdgeCircuitBreakerOptions {
  readonly now?: Date;
}

export async function evaluateEdgeCircuitBreaker(
  kv: KVNamespace,
  request: VerificationRequest,
  options: EdgeCircuitBreakerOptions = {},
): Promise<EdgeCircuitBreakerResult> {
  const nowMs = (options.now ?? new Date()).getTime();
  const sessionId = request.context.session_id ?? request.agent.id;

  const maxRepeats =
    request.authority.constraints.max_repeated_calls ?? DEFAULT_MAX_REPEATED_CALLS;
  const windowSeconds =
    request.authority.constraints.time_window_seconds ?? DEFAULT_TIME_WINDOW_SECONDS;
  const coolOffSeconds =
    request.authority.constraints.cool_off_seconds ?? DEFAULT_COOL_OFF_SECONDS;
  const maxSessionActions =
    request.authority.constraints.max_session_actions ?? DEFAULT_MAX_SESSION_ACTIONS;

  const trippedKey = `cb:tripped:${sessionId}`;
  const epochKey = `cb:epoch:${sessionId}`;

  try {
    // 1. Check if session is currently tripped
    const trippedRaw = await kv.get(trippedKey);
    if (trippedRaw !== null) {
      const trippedData = JSON.parse(trippedRaw) as {
        code: "CIRCUIT_TRIPPED:LOOP_DETECTED" | "CIRCUIT_TRIPPED:BUDGET_EXCEEDED";
        message: string;
        trippedAt: number;
      };
      const elapsedSec = (nowMs - trippedData.trippedAt) / 1000;
      if (elapsedSec < coolOffSeconds) {
        const remaining = Math.max(1, Math.ceil(coolOffSeconds - elapsedSec));
        return {
          tripped: true,
          reasonCode: trippedData.code,
          message: `Circuit breaker active for session '${sessionId}'. Cool-off remaining: ${remaining}s. (${trippedData.message})`,
          details: {
            session_id: sessionId,
            cool_off_remaining_s: remaining,
            reason: trippedData.code,
          },
        };
      }
    }

    // 2. Fetch session epoch (version)
    const epochRaw = await kv.get(epochKey);
    const epoch = epochRaw ?? "0";
    const budgetKey = `cb:budget:${sessionId}:${epoch}`;

    // 3. Check session budget
    const budgetRaw = await kv.get(budgetKey);
    const currentBudget = budgetRaw ? parseInt(budgetRaw, 10) : 0;
    if (currentBudget >= maxSessionActions) {
      const message = `Session action budget of ${maxSessionActions} exceeded for session '${sessionId}'.`;
      await kv.put(
        trippedKey,
        JSON.stringify({
          code: "CIRCUIT_TRIPPED:BUDGET_EXCEEDED",
          message,
          trippedAt: nowMs,
        }),
        { expirationTtl: Math.max(60, Math.ceil(coolOffSeconds)) },
      );
      return {
        tripped: true,
        reasonCode: "CIRCUIT_TRIPPED:BUDGET_EXCEEDED",
        message,
        details: {
          session_id: sessionId,
          total_actions: currentBudget,
          max_session_actions: maxSessionActions,
        },
      };
    }

    // 4. Sliding-window loop detection for identical tool parameters
    const paramHash = await sha256(canonicalize(request.action.parameters));
    const signature = `${request.action.type}:${request.action.target}:${paramHash}`;
    const signatureHash = await sha256(signature);
    const loopKey = `cb:loop:${sessionId}:${epoch}:${signatureHash}`;

    const historyRaw = await kv.get(loopKey);
    const cutoffMs = nowMs - windowSeconds * 1000;
    const timestamps: number[] = historyRaw
      ? (JSON.parse(historyRaw) as number[]).filter((ts) => ts >= cutoffMs)
      : [];

    if (timestamps.length >= maxRepeats) {
      const message = `Potential infinite tool-calling loop detected: action '${request.action.type}' called ${timestamps.length + 1} times with identical parameters within ${windowSeconds}s.`;
      await kv.put(
        trippedKey,
        JSON.stringify({
          code: "CIRCUIT_TRIPPED:LOOP_DETECTED",
          message,
          trippedAt: nowMs,
        }),
        { expirationTtl: Math.max(60, Math.ceil(coolOffSeconds)) },
      );
      return {
        tripped: true,
        reasonCode: "CIRCUIT_TRIPPED:LOOP_DETECTED",
        message,
        details: {
          session_id: sessionId,
          action_type: request.action.type,
          target: request.action.target,
          repeats: timestamps.length + 1,
          window_seconds: windowSeconds,
        },
      };
    }

    // 5. Update history and budget in KV
    timestamps.push(nowMs);
    const loopTtl = Math.max(60, Math.ceil(windowSeconds * 2));
    const budgetTtl = Math.max(3600, Math.ceil(coolOffSeconds * 2));

    await Promise.all([
      kv.put(loopKey, JSON.stringify(timestamps), { expirationTtl: loopTtl }),
      kv.put(budgetKey, String(currentBudget + 1), { expirationTtl: budgetTtl }),
    ]);

    return { tripped: false };
  } catch (error) {
    const errorMsg = error instanceof Error ? error.message : "Unknown KV error";
    console.error(
      JSON.stringify({
        event: "vizier.circuit_breaker.error",
        session_id: sessionId,
        error: errorMsg,
      }),
    );
    // Fail-closed to preserve SECURITY.md invariant 1: if security evaluation errors, block execution
    return {
      tripped: true,
      reasonCode: "CIRCUIT_TRIPPED:LOOP_DETECTED",
      message: `Circuit breaker storage error: ${errorMsg}. Action blocked.`,
      details: {
        session_id: sessionId,
        error: errorMsg,
      },
    };
  }
}

export async function resetEdgeCircuitBreaker(
  kv: KVNamespace,
  sessionId: string,
): Promise<void> {
  await Promise.all([
    kv.delete(`cb:tripped:${sessionId}`),
    kv.put(`cb:epoch:${sessionId}`, String(Date.now()), { expirationTtl: 86400 }),
  ]);
}
